### PR TITLE
o/assertstate: auto aliases for apps that exist

### PR DESCRIPTION
**Pull request by bboozzoo**
Attempt to fix https://bugs.launchpad.net/snapd/+bug/1911016 where snapd would
unnecessarily generate refresh-aliases tasks when there is no real snap update
and no aliases change. The reason is that a snap declaration carries aliases for
apps that are not present in the currently installed snap, thus we always get a
non empty delta of aliases.

### CODE REVIEW

File: overlord/assertstate/assertstate.go [hunk 1]
```diff
@@ -314,7 +314,13 @@ func AutoAliases(s *state.State, info *snap.Info) (map[string]string, error) {
 	}
 	explicitAliases := decl.Aliases()
 	if len(explicitAliases) != 0 {
-		return explicitAliases, nil
+		aliasesForApps := make(map[string]string, len(explicitAliases))
```
Comment (pedronis) [inline overlord/assertstate/assertstate.go:317]: doesn't this break the transition case where higher levels need to be aware to remove aliases to things that don't exist anymore?

Comment (pedronis) [reply overlord/assertstate/assertstate.go:317]: the question might be silly but is the thing we need to be careful about. Relateldy we should check the places that mock this across our unit test and make sure that the mocking respects this new behavior

Comment (bboozzoo) [reply overlord/assertstate/assertstate.go:317]: Our test code did not really cover the scenario of apps going away/coming back while the snap declaration lists a bunch of potentially matching aliases in a realistic way. I've extended the snapstate test a little and added a managers test to make sure that aliases are removed/added.

Comment (pedronis) [reply overlord/assertstate/assertstate.go:317]: I'm still worried that some of our test are not consistent anymore with the changes to AutoAliases in assertstate, if I make this change on this branch:

```
diff --git a/overlord/snapstate/aliasesv2.go b/overlord/snapstate/aliasesv2.go
index 770a18622d..47e9d9195d 100644
--- a/overlord/snapstate/aliasesv2.go
+++ b/overlord/snapstate/aliasesv2.go
@@ -141,6 +141,14 @@ func applyAliasesChange(snapName string, prevAutoDisabled bool, prevAliases map[
        return add, remove, nil
 }
 
+func checkAutoAliases(aliases map[string]string, info *snap.Info) {
+       for _, app := range aliases {
+               if _, ok := info.Apps[app]; !ok {
+                       fmt.Printf("%s missing aliased app %q\n", info.SnapName(), app)
+               }
+       }
+}
+
 // AutoAliases allows to hook support for retrieving the automatic aliases of a snap.
 var AutoAliases func(st *state.State, info *snap.Info) (map[string]string, error)
 
@@ -180,6 +188,7 @@ func autoAliasesDelta(st *state.State, names []string) (changed map[string][]str
                        continue
                }
                autoAliases, err := AutoAliases(st, info)
+               checkAutoAliases(autoAliases, info)
                if err != nil {
                        if firstErr == nil {
                                firstErr = err
@@ -206,6 +215,7 @@ func autoAliasesDelta(st *state.State, names []string) (changed map[string][]str
 // for the snap.
 func refreshAliases(st *state.State, info *snap.Info, curAliases map[string]*AliasTarget) (newAliases map[string]*AliasTarget, err error) {
        autoAliases, err := AutoAliases(st, info)
+       checkAutoAliases(autoAliases, info)
        if err != nil {
                return nil, err
        }

```

I see quite a bit of printing. I haven't studied each test but the differences needs to be fixed, either by adding the app to the snaps or by changing the tests depending what makes sense. But right now we are testing newly incosistent scenarios.

**Review by pedronis — APPROVED** [review]
lgtm

thanks for the new test @bboozzoo 
thanks for the test tweaks @MiguelPires

**Review by MiguelPires — APPROVED** [review]
LGTM

**Pull request merged**